## Info2A-PSI-PSIx_2022-23_TD09_Jeux-stratégies_CR

* extrait du programme

<img src="https://github.com/lbaille20/Info2A-PSIx_2022-23/raw/main/TD09/TD09_BO.png"/>

<img src="https://github.com/lbaille20/Info2A-PSIx_2022-23/raw/main/TD09/TD09_fig01.png"/>

* **énoncé du TD (version 4)** : <a href="https://github.com/lbaille20/Info2A_2022-23/raw/main/Info2A-PSI/TD09/Info2A-PSI-PSIx_TD09_2022-23_Jeux%20Strat%C3%A9gies%20EN_v4.pdf">ici</a>

**Q1.**

**On prêtera attention au fait que le transition possible $4\rightarrow 3$ ne respecte pas le cadre strict du programme***, dans la mesure où elle signifie que le joueur 2, dans l'état 4, a la possibilité d'aller à l'état 3, dans lequel c'est encore à lui de jouer, ce qui ne respecte pas le cadre du programme dans lequel on ne s'intéresse qu'aux jeux dans lesquels les joueurs jouent systématique à tour de rôle.

In [1]:
LS = [i for i in range(7)] ## les sommets sont étiquetés par les entiers de zéro à 6

In [2]:
dC = {0: 1, 1: 2, 2: 1, 3: 2, 4: 2, 5: 1, 6: 2} ## le joueur 1 contrôle les états 0, 2 et 5, le joueur 2 les autres

In [3]:
dAdj = {0: [1, 4], 
        1: [2, 5], 
        2: [3, 6], 
        3: [5, 6], 
        4: [2, 3], 
        5: [], 
        6: []
       }

In [4]:
LObj1, LObj2 = [5], [6] ## il y a deux états finaux, l'état 5 (le joueur 1 gagne), l'état 6 (le joueur 2 gagne)

* définition complète du jeu

In [5]:
jeu = {'etats': LS, 'init': 0, 'controles': dC, 'transitions': dAdj,
       'objectifs1': LObj1, 'objectifs2': LObj2}

In [6]:
jeu

{'etats': [0, 1, 2, 3, 4, 5, 6], 'init': 0, 'controles': {0: 1, 1: 2, 2: 1, 3: 2, 4: 2, 5: 1, 6: 2}, 'transitions': {0: [1, 4], 1: [2, 5], 2: [3, 6], 3: [5, 6], 4: [2, 3], 5: [], 6: []}, 'objectifs1': [5], 'objectifs2': [6]}

**Q2.**

<img src="https://github.com/lbaille20/Info2A-PSIx_2022-23/raw/main/TD09/TD09_fig01.png"/>

Il y a quatre stratégies pour le joueur 1, que voici :

In [7]:
strategie1_1 = {0: 1, 2: 3}; strategie1_2 = {0: 1, 2: 6}

In [8]:
strategie1_3 = {0: 4, 2: 3}; strategie1_4 = {0: 4, 2: 6}

On intègre pas l'état 5, contrôlé par le joueur 1, à ses stratégies, car, si le jeu est dans l'état 5, le joueur 1 a gagné et donc n'aura pas à choisir de coup suivant.

**Q3.**

Il existe 8 stratégies possibles pour le joueur 2, car il contrôle 3 états (les états 1, 3 et 4) et pour chacun d'eux, il peut choisir 2 états suivants.  
On intègre pas l'état 6, contrôlé par le joueur 2, à ses stratégies, car, si le jeu est dans l'état 6, le joueur 2 a gagné et donc n'aura pas à choisir de coup suivant.

L'une de ces stratégies est gagnante, il s'agit de la stratégie suivante :

In [9]:
strategie2_g = {1: 2, 3: 6, 4: 3} # stratégie gagnante pour 2

Quels que soient les coups joués par le joueur 1, à partir de l'état zéro, le joueur 2 est assuré de gagner. Il suffit pour cela :  
- qu'il choisisse l'état 2 s'il a à jouer à partir de l'état 1 ;     
- qu'il choisisse l'état 6 s'il a à jouer à partir de l'état 3 ;     
- qu'il choisisse l'état 3 s'il a à jouer à partir de l'état 4.

**Q4.**

On rappelle la forme de la structure de données représentant un jeu :

In [10]:
jeu = {'etats': LS, 'init': 0, 'controles': dC, 'transitions': dAdj,
       'objectifs1': LObj1, 'objectifs2': LObj2}

afin d'éclairer le code proposé :

In [11]:
def execute(jeu, strategie1, strategie2):
    etat = jeu['init'] ## état courant du jeu
    partie = [etat] ## liste des états successifs au cours de la partie
    while etat not in jeu['objectifs1'] and etat not in jeu['objectifs2']: ## tant qu'aucun joueur n'a atteint un objectif
        joueur_ayant_la_main = jeu['controles'][etat]
        if joueur_ayant_la_main == 1:
            ## vérification de la validité de la stratégie :
            assert strategie1[etat] in dAdj[etat]
            ## application de la stratégie
            etat = strategie1[etat]
        else:
            ## vérification de la validité de la stratégie :
            assert strategie2[etat] in dAdj[etat]
            ## application de la stratégie
            etat = strategie2[etat]
        partie.append(etat)
    return partie

On aurait pu définir une fonction caractérisant la fin de partie :

In [12]:
def fin_partie(jeu, etat):
    return etat in jeu['objectifs1'] or etat in jeu['objectifs2']

et l'utiliser dans l'implémentation de la fonction `execute` :

In [13]:
def execute(jeu, strategie1, strategie2):
    etat = jeu['init'] ## état courant du jeu
    partie = [etat] ## liste des états successifs au cours de la partie
    while not fin_partie(jeu, etat):
        joueur_ayant_la_main = jeu['controles'][etat]
        if joueur_ayant_la_main == 1:
            ## vérification de la validité de la stratégie :
            assert strategie1[etat] in dAdj[etat]
            ## application de la stratégie
            etat = strategie1[etat]
        else:
            ## vérification de la validité de la stratégie :
            assert strategie2[etat] in dAdj[etat]
            ## application de la stratégie
            etat = strategie2[etat]
        partie.append(etat)
    return partie

**Q5.**

<img src="https://github.com/lbaille20/Info2A-PSIx_2022-23/raw/main/TD09/TD09_fig01.png"/>

On reprend le jeu décrit par le dictionnaire calculé en **Q1.**, que l'on a recopié ici :

In [14]:
jeu = {'etats': [0, 1, 2, 3, 4, 5, 6],
       'init': 0,
       'controles': {0: 1, 1: 2, 2: 1, 3: 2, 4: 2, 5: 1, 6: 2},
       'transitions': {0: [1, 4], 1: [2, 5], 2: [3, 6], 3: [5, 6], 4: [2, 3], 5: [], 6: []},
       'objectifs1': [5], 'objectifs2': [6]}

On teste les quatre stratégies possibles pour le joueur 1, contre la stratégie proposée en **Q3.** en tant que stratégie gagnante pour le joueur 2 :

In [15]:
strategie1_1 = {0: 1, 2: 3}
strategie1_2 = {0: 1, 2: 6}
strategie1_3 = {0: 4, 2: 3}
strategie1_4 = {0: 4, 2: 6}

strategie2_g = {1: 2, 3: 6, 4: 3}

for st1 in (strategie1_1, strategie1_2, strategie1_3, strategie1_4):
    print(execute(jeu, st1, strategie2_g))

[0, 1, 2, 3, 6]
[0, 1, 2, 6]
[0, 4, 3, 6]
[0, 4, 3, 6]


On vérifie donc, par un examen exhaustif de toutes les stratégies possibles pour le joueur 1, que la stratégie du joueur 2 est bien gagnante, puisque toutes les parties terminent dans l'état 6.

**Q6.**

On reprend l'algorithme usuel de recherche d'une valeur dans une liste :

In [16]:
def appartientL(e, A):
    for etat in A:
        if etat == e:
            return True
    return False

On rappelle que la complexité de cet algorithme, en fonction du nombre $n$ d'éléments de $A$, est en $O(1)$ dans le meilleur cas (lorsque la valeur cherchée est en première position dans la liste), et en $O(n)$ dans le pire cas (lorsque la valeur cherchée n'est pas dans la liste).

**Q7.**

Si les attracteurs sont définis par des dictionnaires dont les numéros des états appartenant à l'attracteur sont les clés, alors on peut tester l'appartenance d'un état à l'attracteur par un test `e in A` qui est en $O(1)$, c'est-à-dire dont la complexité est indépendante du nombre d'éléments (couples (clé, valeur)) dans le dictionnaire, ce qui est particulièrement intéressant lorsque le nombre d'états du jeu est important (ce qui est le cas pour la plupart des jeux, même très simples : environ 5000 états pour le jeu de morpion étudié dans la suite).

Si les attracteurs sont donnés sous la forme de dictionnaires comme indiqué, la fonction `appartientD` sera implémenté comme suit :

In [17]:
def appartientD(e, A):
    return e in A

<i><u>Note</u> :</i> cette implémentation conviendrait aussi si $A$ est de type `list` mais sa complexité, avec une entrée de type `list` ne serait pas meilleure que celle de la fonction `appartientL`.

**Q8.**

La première fonction examine tous les successeurs possibles de $e$ et renvoie `True` dès qu'un successeur élément de $A$ est trouvé, et `False` si aucun successeur n'est dans $A$ :

In [18]:
def exist_succ(jeu, e, A, appartient):
    succs_e = jeu['transitions'][e]
    for etat in succs_e:
        if appartient(etat, A):
            return True
    return False

La seconde fonction teste si $e$ admet au moins un successeur, et renvoie `False` si ce n'est pas le cas. Si $e$ admet au moins un successeur, la fonction examine tous les successeurs possibles de $e$ et renvoie `False` dès qu'un successeur n'est pas élément de $A$, et `True` s'il s'avère que tous les successeurs sont dans $A$ :

In [19]:
def forall_succ(jeu, e, A, appartient):
    succs_e = jeu['transitions'][e]
    if len(succs_e) == 0:
        return False
    for etat in succs_e:
        if not appartient(etat, A):
            return False
    return True

**Q9.**

On rappelle la forme de la structure de données représentant un jeu afin d'éclairer le code proposé :

In [20]:
jeu = {'etats': LS, 'init': 0, 'controles': dC, 'transitions': dAdj,
       'objectifs1': LObj1, 'objectifs2': LObj2}

Lors d'un appel à `step`, on parcourt l'ensemble des états possibles du jeu, et pour chacun d'eux qui n'est pas déjà dans l'attracteur :
- si l'état est contrôlé par le joueur $j$, on cherche s'il existe un coup lui permettant d'atteindre un état présent dans l'attracteur ;  
- si l'état est contrôlé par l'adversaire du joueur $j$, on cherche si, quelque soit le coup choisi par l'adversaire, le jeu arrive dans un état présent dans l'attracteur.  
On termine en ajoutant à l'attracteur les états vérifiant ces conditions, et en renvoyant un booléen indiquant si de nouveaux états ont été ajoutés à l'attracteur.

In [21]:
def step(jeu, j, A, appartient):
    ## recherche d'états à ajouter à l'attracteur
    nouveaux_etats = []
    for etat in jeu['etats']:
        if not appartientD(etat, A):
            if jeu['controles'][etat] == j:
                if exist_succ(jeu, etat, A, appartientD):
                    nouveaux_etats.append(etat)
                    flag = True
            else:
                if forall_succ(jeu, etat, A, appartientD):
                    nouveaux_etats.append(etat)
                    flag = True
    ## ajouts des nouveaux états à l'attracteur
    for etat in nouveaux_etats:
        A[etat] = None
    ## renvoie du booléen indiquant si de nouveaux états ont été trouvés
    return len(nouveaux_etats) > 0

* tests

<img src="https://github.com/lbaille20/Info2A-PSIx_2022-23/raw/main/TD09/TD09_fig01.png"/>

In [22]:
A = {5: None}
step(jeu, 1, A, appartientD)

False

In [23]:
A = {6: None}
step(jeu, 2, A, appartientD), A

(True, {6: None, 3: None})

In [24]:
A = {6: None}
print(step(jeu, 2, A, appartientD), A)
print(step(jeu, 2, A, appartientD), A)
print(step(jeu, 2, A, appartientD), A)
print(step(jeu, 2, A, appartientD), A)
print(step(jeu, 2, A, appartientD), A)

True {6: None, 3: None}
True {6: None, 3: None, 2: None, 4: None}
True {6: None, 3: None, 2: None, 4: None, 1: None}
True {6: None, 3: None, 2: None, 4: None, 1: None, 0: None}
False {6: None, 3: None, 2: None, 4: None, 1: None, 0: None}


**Q10.**

La fonction consiste à partir d'un attracteur (représenté par un dictionnaire) initialisé avec les états objectifs pour le joueur $j$, et à le compléter par de nouveaux états par appel à la fonction `step`, jusqu'à ce que la suite des états de l'attracteur soit stationnaire *i.e.* tant que les appels à la fonction `step` renvoient `True`.

In [25]:
def attracteur_infty(jeu, j, appartient):
    A = {}
    if j == 1:
        for etat in jeu['objectifs1']:
            A[etat] = None
    else:
        for etat in jeu['objectifs2']:
            A[etat] = None
    ##
    poursuite = True
    while poursuite:
        poursuite = step(jeu, j, A, appartient)
    return [etat for etat in A.keys()]

**Q11.**

* existence d'une stratégie gagnante pour le joueur 2

In [26]:
attracteur_infty(jeu, 2, appartientD)

[6, 3, 2, 4, 1, 0]

L'état initial, 0, est une clé de l'attracteur, donc il existe une stratégie gagnante pour le joueur 2.  
   
Il en découle qu'il n'en existe donc pas pour le joueur 1, ce que l'on vérifie ci-après.

* existence d'une stratégie gagnante pour le joueur 1

In [27]:
attracteur_infty(jeu, 1, appartientD)

[5]

On constate bien que l'état initial, 0, n'est pas dans les clés de l'attracteur pour le joueur 1.

**Q12.**

On modifie la fonction `exist_succ` en la fonction `exist_succ_strategie` de sorte que, lorsque pour un état $e$, un successeur est trouvé, une clé soit ajoutée pour l'état $e$ dans le dictionnaire `strategie`, associée à ce successeur.

In [28]:
def exist_succ_strategie(jeu, e, A, appartient, strategie):
    succs_e = jeu['transitions'][e]
    for etat in succs_e:
        if appartient(etat, A):
            ### modification ICI ###
            strategie[e] = etat 
            ### modification FIN ###
            return True
    return False

In [29]:
def step_strategie(jeu, j, A, appartient, strategie):
    ## recherche d'états à ajouter à l'attracteur
    nouveaux_etats = []
    for etat in jeu['etats']:
        if not appartient(etat, A):
            if jeu['controles'][etat] == j:
                ### modification ICI ###
                if exist_succ_strategie(jeu, etat, A, appartient, strategie):
                    ### modification FIN ###
                    nouveaux_etats.append(etat)
                    flag = True
            else:
                if forall_succ(jeu, etat, A, appartient):
                    nouveaux_etats.append(etat)
                    flag = True
    ## ajouts des nouveaux états à l'attracteur
    for etat in nouveaux_etats:
        A[etat] = None
    ## renvoie du booléen indiquant si de nouveaux états ont été trouvés
    return len(nouveaux_etats) > 0

In [30]:
def attracteur_infty_strategie(jeu, j, appartient):
    strategie = {}
    A = {}
    if j == 1:
        for etat in jeu['objectifs1']:
            A[etat] = None
    else:
        for etat in jeu['objectifs2']:
            A[etat] = None
    ##
    poursuite = True
    while poursuite:
        poursuite = step_strategie(jeu, j, A, appartient, strategie)
    return strategie

* test

<img src="https://github.com/lbaille20/Info2A-PSIx_2022-23/raw/main/TD09/TD09_fig01.png"/>

In [31]:
attracteur_infty_strategie(jeu, 2, appartientD)

{3: 6, 4: 3, 1: 2}

On retrouve la stratégie gagnante pour le joueur 2 trouvée précédemment.

In [32]:
attracteur_infty_strategie(jeu, 1, appartientD)

{}

On constate qu'aucun état ne permet au joueur 1 d'obtenir la victoire.

## Jeu de morpion

<img src="https://github.com/lbaille20/Info2A-PSIx_2022-23/raw/main/TD09/TD09_fig02.png"/>

In [33]:
symboles = {0: '_', 1: 'O', 2: 'X'}

**Q13.**

In [34]:
def plateau_vide():
    return [['_'] * 3 for i in range(3)]

In [35]:
plateau_vide()

[['_', '_', '_'], ['_', '_', '_'], ['_', '_', '_']]

On donne la fonction `affiche` permettant d'afficher agréablement une configuration du plateau de jeu :

In [36]:
symboles = {0: '_', 1: 'O', 2: 'X'}

def affiche(plateau):
    for ligne in plateau:
        print(' '.join(ligne))
        
## test
P0 = plateau_vide()
affiche(P0)

_ _ _
_ _ _
_ _ _


**Q14.**

In [37]:
def jouer(plateau, i, j, k):
    plateau[i][j] = symboles[k]

On pourrait compléter la fonction en vérifiant au titre de précondition que la case à remplir est bien vide :

In [38]:
def jouer(plateau, i, j, k):
    assert plateau[i][j] == symboles[0]
    plateau[i][j] = symboles[k]

* tests

In [39]:
P = plateau_vide()
jouer(P, 0, 0, 1)
affiche(P)

O _ _
_ _ _
_ _ _


In [40]:
P = plateau_vide()
jouer(P, 0, 0, 2)
affiche(P)

X _ _
_ _ _
_ _ _


In [41]:
P = plateau_vide()
jouer(P, 0, 0, 1)
affiche(P)

O _ _
_ _ _
_ _ _


Si maintenant on essaie de faire joueur le joueur 2 sur la même case :

In [42]:
affiche(P)
jouer(P, 0, 0, 2)

O _ _
_ _ _
_ _ _


Traceback (most recent call last):
  File "<input>", line 2, in <module>
  File "<input>", line 2, in jouer
AssertionError


Error: 

**Q15.**

In [43]:
def lire(plateau, i, j):
    for k in range(3):
        if plateau[i][j] == symboles[k]:
            return k

* tests

In [44]:
P = [['_', 'X', '_'], ['O', '_', '_'], ['_', '_', '_']]
print(lire(P, 0, 1), lire(P, 1, 0), lire(P, 2, 2))

2 1 0


**Q16.**

La fonction `controle(conf)` vérifie que le nombre de cellules occupées par le joueur 1, $n_1$, ou le joueur 2, $n_2$, sont soit égaux (c'est alors au joueur 1 et jouer), ou que $n_1=n_2+1$ (c'est alors au joueur 2 de jouer).

In [45]:
def controle(conf):
    cpt1, cpt2 = 0, 0
    for i in range(3):
        for j in range(3):
            if conf[i][j] == symboles[1]:
                cpt1 += 1
            elif conf[i][j] == symboles[2]:
                cpt2 += 1
    assert 0 <= cpt1 - cpt2 and cpt1 - cpt2 <= 1
    if cpt1 - cpt2 == 0:
        return 1
    else:
        return 2

* tests

In [46]:
P = plateau_vide()
controle(P)

1

In [47]:
controle([['_', 'X', '_'], ['O', '_', '_'], ['_', '_', '_']])

1

In [48]:
controle([['_', 'X', '_'], ['O', '_', '_'], ['_', 'O', '_']])

2

In [49]:
controle([['_', 'X', '_'], ['O', '_', '_'], ['_', 'X', '_']])

Traceback (most recent call last):
  File "<input>", line 1, in <module>
  File "<input>", line 9, in controle
AssertionError


Error: 

**Q17.**

Dans l'implémentation proposée, pour chacun des huit alignements possibles (en ligne, en colonne ou selon les diagonales), on fait la liste des symboles présents pour la comparer une liste des trois symboles du joueur $k$ (sur l'exemple : soit à la liste `['O', 'O', 'O']`, pour le joueur 1, soit à la liste  `['X', 'X', 'X']`, pour le joueur 2).

In [50]:
def gagnante(conf, k):
    G = {1: [symboles[1]] * 3, 2: [symboles[2]] * 3} # G[1] = ['O', 'O', 'O'], G[2] = ['X', 'X', 'X'] ici
    ## alignements horizontaux
    for i in range(3):
        L = []
        for j in range(3):
            L.append(conf[i][j])
        # L est la liste des symboles sur la ligne i
        if L == G[k]:
            return True
    ## alignements verticaux
    for j in range(3):
        L = []
        for j in range(3):
            L.append(conf[i][j])
        # L est la liste des symboles sur la colonne j
        if L == G[k]:
            return True
    ## alignements diagonaux
    L = []
    for i in range(3):
        L.append(conf[i][i])
    # L est la liste des symboles sur la diagonale descendante
    if L == G[k]:
        return True
    L = []
    for i in range(3):
        L.append(conf[i][2 - i])
    # L est la liste des symboles sur la diagonale ascendante
    if L == G[k]:
        return True
    return False

En utilisant des compréhensions de listes :

In [51]:
def gagnante(conf, k):
    G = {1: [symboles[1]] * 3, 2: [symboles[2]] * 3} # G[1] = ['O', 'O', 'O'], G[2] = ['X', 'X', 'X'] ici
    ## alignements horizontaux
    for i in range(3):
        if [conf[i][j] for j in range(3)] == G[k]:# examen de la liste des symboles sur la ligne i
            return True
    ## alignements verticaux
    for j in range(3):
        if [conf[i][j] for i in range(3)] == G[k]:# examen de la liste des symboles sur la colonne j
            return True
    ## alignements diagonaux
    if [conf[i][i] for i in range(3)] == G[k]:# examen de la liste des symboles sur la la diagonale descendante
        return True
    if [conf[i][2 - i] for i in range(3)] == G[k]:# examen de la liste des symboles sur la la diagonale ascendante
        return True
    ## si aucun alignement de trois symbole n'a été trouvé
    return False

* tests

In [52]:
conf = [['_', 'X', '_'], ['O', '_', '_'], ['_', 'O', '_']]
controle(conf)
affiche(conf)
gagnante(conf, 1), gagnante(conf, 2)

_ X _
O _ _
_ O _


(False, False)

In [53]:
conf = [['_', 'X', 'O'], ['X', 'O', '_'], ['O', 'X', '_']]
controle(conf)
affiche(conf)
gagnante(conf, 1), gagnante(conf, 2)

_ X O
X O _
O X _


(True, False)

In [54]:
conf = [['_', 'X', 'O'], ['X', 'X', '_'], ['O', 'O', 'O']]
controle(conf)
affiche(conf)
gagnante(conf, 1), gagnante(conf, 2)

_ X O
X X _
O O O


(True, False)

In [55]:
conf = [['_', 'X', 'O'], ['_', 'X', '_'], ['O', 'X', '_']]
affiche(conf)
gagnante(conf, 1), gagnante(conf, 2)

_ X O
_ X _
O X _


(False, True)

### 3.1. Première approche pour la construction d’une représentation du jeu

**Q18.**

Le plateau comporte neuf cases et, avec les conditions spécifiées, chacune de ces neuf cases peut, soit être vide, soit occupée par le joueur 1, soit par le joueur 2.  
Une configuration peut donc être représentée par un 9-uplet quelconque de valeurs prises dans $[\![0;2]\!]$ (un 0 représentant une case vide, 1 et 2 une case occupée soit par le joueur 1, soit par le joueur 2).
Il y a donc $\left|[\![0;2]\!]^9\right|=\left|[\![0;2]\!]\right|^9=3^9$ configurations possibles dans ces conditions.

In [56]:
3 ** 9

19683

**Q19.** Génération de tous les états possibles du plateau

In [57]:
from copy import deepcopy

symboles = {0: '_', 1: 'O', 2: 'X'}

confs_all = [plateau_vide()]
for i in range(3):
    for j in range(3):
        new_confs = []
        for conf in confs_all:
            for k in range(2):
                new_conf = deepcopy(conf)
                jouer(new_conf, i, j, k + 1)
                new_confs.append(new_conf)
            #print(new_confs)
        confs_all.extend(new_confs)

print(len(confs_all))

19683


* examen d'une configuration prise au hasard parmi les configurations générées

In [58]:
from random import randrange

no = randrange(len(confs_all))
print(no)
affiche(confs_all[no])

11651
O _ O
O O _
_ O O


**Q20.**

La fonction teste si le nombre de cases occupées par le joueur 1 est égal au nombre de cases occupées par le joueur 2, ou lui est supérieur de 1 uniquement.

In [59]:
def verif1(conf):
    cpt1, cpt2 = 0, 0
    for i in range(3):
        for j in range(3):
            if conf[i][j] == symboles[1]:
                cpt1 += 1
            elif conf[i][j] == symboles[2]:
                cpt2 += 1
    return 0 <= cpt1 - cpt2 and cpt1 - cpt2 <= 1

* tests

In [60]:
P = [['O', 'O', 'O'], ['X', 'O', 'X'], ['O', 'X', 'X']]
affiche(P)
verif1(P)

O O O
X O X
O X X


True

In [61]:
P = [['_', 'X', '_'], ['O', '_', '_'], ['_', 'O', '_']]
affiche(P)
verif1(P)

_ X _
O _ _
_ O _


True

In [62]:
P = [['_', 'X', '_'], ['O', '_', '_'], ['_', 'X', '_']]
affiche(P)
verif1(P)

_ X _
O _ _
_ X _


False

**Q21.**

La fonction teste qu'il n'existe aucun alignement de trois symboles pour le joueur controlant la configuration.

In [63]:
def verif2(conf):
    G = {1: [symboles[1]] * 3, 2: [symboles[2]] * 3} # G = {1: ['O', 'O', 'O'], 2: ['X', 'X', 'X']} ici
    joueur = controle(conf) ## joueur contrôlant la configuration
    ## alignements horizontaux
    for i in range(3):
        if [conf[i][j] for j in range(3)] == G[joueur]:# examen de la liste des symboles sur la ligne i
            return False ## rejet de la configuration si le joueur ayant le contrôle a un alignement gagnant
    ## alignements verticaux
    for j in range(3):
        if [conf[i][j] for i in range(3)] == G[joueur]:# examen de la liste des symboles sur la colonne j
            return False ## rejet de la configuration si le joueur ayant le contrôle a un alignement gagnant
    ## alignements diagonaux
    if [conf[i][i] for i in range(3)] == G[joueur]:# examen de la liste des symboles sur la la diagonale descendante
        return False ## rejet de la configuration si le joueur ayant le contrôle a un alignement gagnant
    if [conf[i][2 - i] for i in range(3)] == G[joueur]:# examen de la liste des symboles sur la la diagonale ascendante
        return False ## rejet de la configuration si le joueur ayant le contrôle a un alignement gagnant
    return True 

* tests

In [64]:
P = [['O', 'O', 'O'], ['X', 'O', 'X'], ['O', 'X', 'X']]
affiche(P)
verif2(P)

O O O
X O X
O X X


True

In [65]:
P = [['O', 'X', '_'], ['O', '_', '_'], ['_', 'X', 'O']]
affiche(P)
verif2(P)

O X _
O _ _
_ X O


True

In [66]:
P = [['O', 'X', '_'], ['O', 'X', '_'], ['_', 'X', 'O']]
affiche(P)
verif2(P)

O X _
O X _
_ X O


True

In [67]:
P = [['O', 'X', 'X'], ['O', '_', '_'], ['O', 'X', 'O']]
affiche(P)
verif2(P)

O X X
O _ _
O X O


True

In [68]:
P = [['O', 'X', 'X'], ['O', 'O', 'X'], ['O', 'X', 'O']]
affiche(P)
verif2(P)

O X X
O O X
O X O


True

In [69]:
P = [['X', 'X', 'X'], ['O', 'O', 'O'], ['O', 'X', '_']]
affiche(P)
verif2(P)

X X X
O O O
O X _


False

In [70]:
P = [['O', 'X', 'X'], ['O', 'O', 'O'], ['O', 'X', 'X']]
affiche(P)
verif2(P)

O X X
O O O
O X X


True

In [71]:
P = [['O', 'O', 'X'], ['O', 'O', 'X'], ['O', 'X', 'X']]
affiche(P)
verif2(P)

O O X
O O X
O X X


False

**Q22.**

On réduit ici la liste de toutes les configurations possibles du plateau de jeu, à une liste `confs1` des seules configurations vérifiant les conditions testées en questions **Q20** et **Q21**.

In [72]:
confs1 = []
for conf in confs_all:
    if verif1(conf) and verif2(conf):
        confs1.append(conf)

len(confs_all), len(confs1)

(19683, 5478)

* examen d'une configuration prise au hasard parmi les configurations retenues

In [73]:
from random import randrange

no = randrange(len(confs))
print(no)
affiche(confs[no])

Traceback (most recent call last):
  File "<input>", line 3, in <module>
NameError: name 'confs' is not defined


Error: 

* Définition de la liste des numéros d'états du jeu

In [74]:
LS1 = [i for i in range(len(confs1))]
len(LS1)

5478

**Q23.**

* dictionnaire des états contrôlés par l'un ou l'autre joueur

In [75]:
dC1 = {}
for no in LS1:
    joueur = controle(confs1[no])
    dC1[no] = joueur

print(len(LS1), len(confs1), len(dC1))

5478 5478 5478


On peut examiner un état, choisi au hasard :

In [76]:
from random import randrange

etat = randrange(len(dC1))
print(etat, dC1[etat])
affiche(confs1[etat])

5344 1
X _ O
X O X
X O O


* listes des états objectifs (gagnants) pour l'un ou l'autre joueur

In [77]:
LObj11, LObj21 = [], []
for no in LS1:
    if gagnante(confs1[no], 1):
        LObj11.append(no)
    if gagnante(confs1[no], 2):
        LObj21.append(no)

len(LObj11), len(LObj21), len(LObj11) + len(LObj21)

(626, 316, 942)

On peut examiner quelques-uns des états objectifs, choisis au hasard :

In [78]:
from random import randrange

i = randrange(len(LObj11))
etat = LObj11[i]
print(etat)
affiche(confs1[etat])

1971
_ X O
X O X
O O _


In [79]:
from random import randrange

i = randrange(len(LObj21))
etat = LObj21[i]
print(etat)
affiche(confs1[etat])

2191
X X X
O O _
_ _ O


**Q24.**

In [80]:
def transition(conf1, conf2):
    joueur1 = controle(conf1) 
    cpt = 0 # compteur du nombre de cases de contenu différent
    for i in range(3):
        for j in range(3):
            if conf1[i][j] != conf2[i][j]:
                if cpt == 1: ## si on a déjà trouvé une case de contenu différent
                    return False
                ## s'il s'agit de la première case de contenu différent trouvé
                if conf1[i][j] != '_' or conf2[i][j] != symboles[joueur1]:
                    return False
                cpt = 1
    # si le parcours simultané des deux tableaux termine sans que False ait été renvoyé, 
    # il reste à vérifier qu'il y a bien une case de contenu différent et non aucune
    return cpt == 1

* tests

On peut, entre deux configurations entre lesquelles on sait qu'une transition est possible, vérifier que la fonction `transition` renvoie bien `True` :

In [81]:
conf1 = [['X', 'X', '_'], ['O', '_', 'O'], ['_', '_', '_']]
affiche(conf1)
print("*****")
conf2 = [['X', 'X', '_'], ['O', '_', 'O'], ['_', 'O', '_']]
affiche(conf2)
print(transition(conf1, conf2))

X X _
O _ O
_ _ _
*****
X X _
O _ O
_ O _
True


On peut rechercher les états vers lesquels est possible une transition depuis un état choisi aléatoirement :

In [82]:
from random import randrange

## on choisit au hasard un état contrôlé par le joueur 1
etat1 = LS1[randrange(len(LS1))]

print(etat1, controle(confs1[etat1]), confs1[etat1])
affiche(confs1[etat1])
print("*****")

## recherche des états vers lesquels une transition est possible depuis l'état 1
L = []
for etat2 in LS1:
    if transition(confs1[etat1], confs1[etat2]):
        L.append(etat2)
## affichage des résultats
print(L)
for etat in L:
    affiche(confs1[etat])
    print("*****")

1166 1 [['_', 'X', 'X'], ['O', 'X', 'O'], ['_', 'O', '_']]
_ X X
O X O
_ O _
*****
[1184, 1976, 4072]
O X X
O X O
_ O _
*****
_ X X
O X O
O O _
*****
_ X X
O X O
_ O O
*****


**Q25.**

L'algorithme élémentaire pour construire le dictionnaire `dAdj1` des listes d'adjacence de chaque état, consiste à rechercher, pour chaque état du jeu, parmi tous les états du jeu, ceux vers lesquels une transition est possible, en testant l'existence d'une transition possible par appel à la fonction `transition`.  
On met pour cela en place une imbrication de deux boucles parcourant toutes deux tous les états possibles.  
Comme il y a 5478 états possibles, l'algorithme générera $5478\times 5478$ appels à la fonction `transition`, soit plus de 25 millions d'appels.  
  
On peut vérifier en décommentant et en exécutant la cellule ci-dessous, qu'il faut environ deux minutes quarante (160 s) pour construire le dictionnaire `dAdj1`.  
On mesure ici le temps d'exécution en faisant appel à la fonction `time` du module `time`, renvoyant l'heure actuelle en secondes (flottant égal au nombre de secondes écoulées depuis le 1er janvier 1970).

In [83]:
from time import time

t0 = time() ## enregistrement du temps avant exécution

"""
dAdj1 = {}
for etat1 in LS1:
    dAdj1[etat1] = []
    conf1 = confs1[etat1]
    joueur = controle(conf1) ## joueur contrôlant l'état1
    adversaire = 3 - joueur  ## calcule du numéro de l'adversaire du joueur contrôlant l'état1
    for etat2 in LS1:
        conf2 = confs1[etat2]
        if transition(conf1, conf2):
            dAdj1[etat1].append(etat2)
"""
                
t = time() - t0 ## calcul de la durée d'exécution
print(t)        ## affichage de la durée d'exécution (exemple de durée obtenue : 157.92499995231628 s)

0.0


In [84]:
#len(dAdj1), sum([len(dAdj1[etat]) for etat in dAdj1.keys()]) ## ordre et taille du graphe : (5478, 16167) 

* première amélioration du temps de calcul

On peut limiter les appels à la fonction `transition`aux cas où les deux configurations testées sont contrôlées par deux joueurs différents (ce qui doit être dans un déroulement normal du jeu, puisque les joueurs jouent à tour de rôle).  
On peut vérifier en décommentant et en exécutant la cellule ci-dessous, qu'il faut alors environ moitié moins de temps pour construire le dictionnaire `dAdj1` (environ 80 s).  

In [85]:
from time import time

t0 = time() ## enregistrement du temps avant exécution

"""
dAdj1 = {}
for etat1 in LS1:
    dAdj1[etat1] = []
    conf1 = confs1[etat1]
    joueur = controle(conf1) ## joueur contrôlant l'état1
    adversaire = 3 - joueur  ## calcule du numéro de l'adversaire du joueur contrôlant l'état1
    for etat2 in LS1:
        if dC1[etat2] == adversaire: ## on teste que l'état 2 soit contrôlé par l'adversaire
            conf2 = confs1[etat2]
            if transition(conf1, conf2):
                dAdj1[etat1].append(etat2)
"""

t = time() - t0 ## calcul de la durée d'exécution
print(t)        ## affichage de la durée d'exécution (exemple de durée obtenue : 82.53600001335144 s)

0.0


In [86]:
#len(dAdj1), sum([len(dAdj1[etat]) for etat in dAdj1.keys()]) ## ordre et taille du graphe : (5478, 16167) 

* deuxième amélioration du temps de calcul

On peut de plus limiter les appels à la fonction `transition` aux cas où les nombre de cases occupées dans les deux configurations ne diffère que de 1, avec une case occupée de plus dans la deuxième configuration (ce qui traduit qu'entre un état et son suivant, un unique coup a été joué).  
Pour cela on peut définir une fonction, `nb_occcupees`, calculant le nombre de cases occupées dans une configuration : 

In [87]:
def nb_occupees(conf):
    cpt = 0
    for i in range(3):
        for j in range(3):
            if conf[i][j] != symboles[0]:
                cpt += 1
    return cpt

## test
nb_occupees([['O', 'O', '_'], ['X', '_', 'O'], ['_', 'X', '_']])

5

puis construire un dictionnaire, `Dconfs1`, dont les clés sont les valeurs possibles du nombre de cases occupées (de zéro à 9), associées aux numéros des états qui comportent un tel nombre de cases occupées (note : pour l'utilisation suivante du dictionnaire, il sera utile de rajouter la clé 10, associée à la liste vide) :

In [88]:
Dconfs1 = {}
## initialisation du dictionnaire
for nb in range(11):
    Dconfs1[nb] = []
## complétion du dictionnaire
for etat in LS1:
    n = nb_occupees(confs1[etat])
    Dconfs1[n].append(etat)

## vérification
sum([len(Dconfs1[n]) for n in range(11)])

5478

On peut alors vérifier en exécutant la cellule ci-dessous, que le temps de calcul pour construire le dictionnaire `dAdj1` est approximativement divisé par quatre par rapport au précédent (environ 25 s). 

In [89]:
from time import time

t0 = time()

dAdj1 = {}
for etat1 in LS1:
    dAdj1[etat1] = []
    conf1 = confs1[etat1]
    n = nb_occupees(conf1)
    for etat2 in Dconfs1[n + 1]: ## <- ici l'utilité d'avoir définie une clé égale à 10 associée à une liste vide
        conf2 = confs1[etat2]
        if transition(conf1, conf2):
            dAdj1[etat1].append(etat2)

t = time() - t0
print(t) # (exemple de durée obtenue : 25.908999919891357 s)

26.229000329971313


On est ainsi parvenu à gagner un facteur 6 ou 7 sur le temps d'exécution.

* quelques vérifications

In [90]:
len(dAdj1), sum([len(dAdj1[etat]) for etat in dAdj1.keys()]) ## ordre et taille du graphe : (5478, 16167) 

(5478, 16167)

In [91]:
dAdj1[0] ## états suivants possibles à partir l'état initial (plateau vide)

[1, 2, 5, 13, 35, 96, 265, 730, 2023]

In [92]:
etat = 0
for etat in dAdj1[etat]:
    affiche(confs1[etat])
    print("*****")

O _ _
_ _ _
_ _ _
*****
_ O _
_ _ _
_ _ _
*****
_ _ O
_ _ _
_ _ _
*****
_ _ _
O _ _
_ _ _
*****
_ _ _
_ O _
_ _ _
*****
_ _ _
_ _ O
_ _ _
*****
_ _ _
_ _ _
O _ _
*****
_ _ _
_ _ _
_ O _
*****
_ _ _
_ _ _
_ _ O
*****


On peut vérifier que tous les états sont atteignables :

In [93]:
atteignables = [0]
for etat, Ladj in dAdj1.items():
    for e in Ladj:
        if e not in atteignables:
            atteignables.append(e)
len(atteignables), len(confs1)

(5478, 5478)

On peut examiner les états suivants possibles (successeurs) à partir d'un état choisi aléatoirement :

In [94]:
from random import randrange

etat1 = randrange(len(LS1))
print(confs1[etat1])
affiche(confs1[etat1])

suivants = dAdj1[etat1]
print(len(suivants), "successeur(s) :")
for etat2 in suivants:
    affiche(confs1[etat2])
    print("*****")

[['_', 'O', '_'], ['X', '_', 'O'], ['X', 'O', 'X']]
_ O _
X _ O
X O X
3 successeur(s) :
O O _
X _ O
X O X
*****
_ O O
X _ O
X O X
*****
_ O _
X O O
X O X
*****


* test

On peut simuler une partie dans laquelle les deux joueurs jouent aléatoirement à chacun leur tour, un coup parmi tous les coups possibles :

In [95]:
etat = 0
affiche(confs1[etat])
print("\n***** état atteint :", etat)
etats_suivants_possibles = dAdj1[etat]
while len(etats_suivants_possibles) > 0:
    i = randrange(len(etats_suivants_possibles)) # choix du coup suivant
    etat = etats_suivants_possibles[i]
    affiche(confs1[etat])
    print("\n***** état atteint :", etat)
    etats_suivants_possibles = dAdj1[etat]

_ _ _
_ _ _
_ _ _

***** état atteint : 0
_ O _
_ _ _
_ _ _

***** état atteint : 2
_ O _
_ _ _
_ X _

***** état atteint : 733
_ O _
_ _ _
O X _

***** état atteint : 1205
X O _
_ _ _
O X _

***** état atteint : 1211
X O _
_ _ O
O X _

***** état atteint : 1518
X O _
_ X O
O X _

***** état atteint : 1732
X O _
O X O
O X _

***** état atteint : 1885
X O _
O X O
O X X

***** état atteint : 5263


* construction de la représentation du jeu

In [96]:
jeu1 = {'etats': LS1, 'init': 0, 'controles': dC1, 'transitions': dAdj1,
       'objectifs1': LObj11, 'objectifs2': LObj21}

On peut enregistrer les différents résultats dans des fichiers :

In [97]:
f = open("confs_LS1.csv", "w")
for no in LS1:
    #f.write(str(confs[no] + '\t' + str(no) + '\n'))
    f.write(';'.join([s for line in confs1[no] for s in line]) + ';' + str(no) + '\n')
f.close()

In [98]:
f = open("confs_dC1.csv", "w")
for no, joueur in dC1.items():
    f.write(';'.join((str(no), str(joueur))) + '\n')
f.close()

In [99]:
f = open("confs_dAdj1.csv", "w")
for etat, Ladj in dAdj1.items():
    #f.write(str(confs[no] + '\t' + str(no) + '\n'))
    f.write(';'.join([str(etat)] + [str(s) for s in Ladj])+ '\n')
f.close()

In [100]:
f = open("confs_LObj11.csv", "w")
for e in LObj11:
    f.write(str(e) + '\n')
f.close()

In [101]:
f = open("confs_LObj21.csv", "w")
for e in LObj21:
    f.write(str(e) + '\n')
f.close()

### 3.2. Deuxième approche pour la construction d’une représentation du jeu

Dans cette approche, on obtient toutes les configurations possibles en exploration le graphe du jeu à partir de la configuration initiale correspondant à un plateau vide.

On propose pour cela de procéder un parcours en largeur du graphe, en utilisant une file.  

* utilisation d'une structure de données de type `deque` pour modéliser une file FIFO

Les structures de données de type `deque` du module `collections` permettent de modéliser des files de type FIFO (*First In First Out*), ***en proposant les quatre opérations de base requises*** (créer un file vide, tester si une file est vide, enfiler un élément, défiler un élément) ***avec la complexité constante recherchée*** (complexité indépendante de la taille de la file, *i.e.* indépendante du nombre d'éléments dans la file).

Dans la cellules suivante, on peut lire la documentation Python sur les objets de type `deque` :

In [102]:
from collections import deque

help(deque)

Help on class deque in module collections:

class deque(builtins.object)
 |  deque([iterable[, maxlen]]) --> deque object
 |  
 |  A list-like sequence optimized for data accesses near its endpoints.
 |  
 |  Methods defined here:
 |  
 |  __add__(self, value, /)
 |      Return self+value.
 |  
 |  __bool__(self, /)
 |      self != 0
 |  
 |  __contains__(self, key, /)
 |      Return key in self.
 |  
 |  __copy__(...)
 |      Return a shallow copy of a deque.
 |  
 |  __delitem__(self, key, /)
 |      Delete self[key].
 |  
 |  __eq__(self, value, /)
 |      Return self==value.
 |  
 |  __ge__(self, value, /)
 |      Return self>=value.
 |  
 |  __getattribute__(self, name, /)
 |      Return getattr(self, name).
 |  
 |  __getitem__(self, key, /)
 |      Return self[key].
 |  
 |  __gt__(self, value, /)
 |      Return self>value.
 |  
 |  __iadd__(self, value, /)
 |      Implement self+=value.
 |  
 |  __imul__(self, value, /)
 |      Implement self*=value.
 |  
 |  __init__(self, /, 

On crée une file vide avec l'instruction suivante :

In [103]:
f = deque()

On teste si une file est vide en examinant si son nombre d'éléments (donné par la fonction `len`) est égal à zéro :

In [104]:
f = deque()
len(f) == 0

True

On ajoute (on enfile) un élément dans la file à l'aide de la méthode `.appendleft` :

In [105]:
f = deque()
print(f)
f.appendleft(1)
f.appendleft(2)
print(f)
for i in range(3, 11):
    f.appendleft(i)
print(f)

deque([])
deque([2, 1])
deque([10, 9, 8, 7, 6, 5, 4, 3, 2, 1])


On retire un élément de la file (on défile) à l'aide de la méthode `.pop` :

In [106]:
for i in range(10):
    x = f.pop()
    print(f, x)

deque([10, 9, 8, 7, 6, 5, 4, 3, 2]) 1
deque([10, 9, 8, 7, 6, 5, 4, 3]) 2
deque([10, 9, 8, 7, 6, 5, 4]) 3
deque([10, 9, 8, 7, 6, 5]) 4
deque([10, 9, 8, 7, 6]) 5
deque([10, 9, 8, 7]) 6
deque([10, 9, 8]) 7
deque([10, 9]) 8
deque([10]) 9
deque([]) 10


On observe que les éléments ont bien été retirés de la file selon le principe &laquo; *premier entré, premier sorti* &raquo;.

Notons enfin que chercher à défiler un élement sur une file vide déclenche une erreur :

In [107]:
f = deque()
f.pop()

Traceback (most recent call last):
  File "<input>", line 2, in <module>
IndexError: pop from an empty deque


Error: 

**Q27.**

On propose ci-dessous une implémentation de l'algorithme décrit dans l'énoncé, tout d'abord sans utiliser de dictionnaire des configurations, puis en construisant un dictionnaire, `Dconfs2` des configurations, ce qui permet de rechercher avec une complexité constante si une configuration a déjà été obtenue, et permet d'obtenir une représentation du jeu en diminuant le temps d'exécution d'un facteur 10. 

* sans dictionnaire des configurations

In [108]:
from time import time

t0 = time()

from collections import deque
from copy import deepcopy

## initialisations
confs2 = [plateau_vide()] ## liste des configurations possibles
conf_init = 0             ## numéro de la configuration initiale
LS2 = [0]                 ## liste des numéros d’états (position de la configuration dans confs2)
dC2 = {0: 1}              ## dictionnaire indiquant pour chaque état le joueur qui le contrôle
dAdj2 = {0: []}           ## dictionnaire donnant les états atteignables depuis chaque état
LObj12, LObj22 = [], []   ## listes des états objectifs pour les joueurs 1 et 2

## initialisation de la file
file = deque()
file.appendleft(0)
## parcours en largeur du graphe des états du jeu
while len(file) > 0:
    # on défile un numéro d'état (ce numéro donne la position de la configuration correspondante dans la liste confs2)
    conf_no = file.pop()
    # récupération de la configuration correspondante
    conf = confs2[conf_no]
    # récupération du numéro du joueur contrôlant la configuration
    joueur = dC2[conf_no]
    ## pour chaque case vide dans la configuration défilée, on examine toutes les configurations possibles au coup suivant 
    ## pour cela on parcourt la grille à l'aide de deux boucles imbriquées
    for i in range(3):
        for j in range(3):
            ## si une case est vide :
            if lire(conf, i, j) == 0:
                ## on en crée une copie profonde
                next_conf = deepcopy(conf) ## ou bien next_conf =  [[conf[i][j] for j in range(3)] for i in range(3)]
                ## on fait occuper la case vide par le joueur contrôlant la confinguration
                jouer(next_conf, i, j, joueur)
                ## on doit maintenant examiner si la configuration construite, next_conf, a déjà été trouvée ou non :
                ##  - si elle a déjà été atteinte, on doit récupérer le numéro d'état correspondant, i.e. sa position
                ##    dans la liste confs2 ;
                ##  - si elle n'a jamais été atteinte, on doit l'ajouter à la liste confs2, et actualiser la liste
                ##    des numéros d'états, ainsi que le dictionnaire dC2, et, si elle est gagnante, l'ajouter à LObj12
                ##    ou LObj22.
                
                ## On initialise le numéro de la configuration à la valeur len(confs2), qui sera sa position dans 
                ## la liste confs2 si on doit l'y ajouter
                next_conf_no = len(confs2)
                ## puis on parcourt la liste confs2 : si on trouve la configuration, on donne à next_conf_no
                ## la valeur de la position de la configuration dans la liste confs2
                for k in range(len(confs2)):
                    if next_conf == confs2[k]:
                        next_conf_no = k ## <- si la configuration n'est pas nouvelle,
                                         ## son numéro est donné par sa position dans la liste confs2
                        break ## on termine la boucle si la configuration a été trouvée 
                ## on examine si la configuration a été trouvée en examinant si next_conf_no a toujours pour valeur
                ## len(confs2) ou non, si tel est le cas, il s'agit d'une nouvelle configuration et on doit l'ajouter
                ## à la fin de la liste confs2, et procéder aux actualisations nécessaires
                if next_conf_no == len(confs2):
                    confs2.append(next_conf)
                    ## on ajoute le numéro de cette nouvelle configuration à la liste des numéros d'état
                    LS2.append(next_conf_no)
                    ## on compléte le dictionnaire dC2 pour indiquer quel joueur contrôle cet état
                    dC2[next_conf_no] = 3 - joueur ## 3 - joueur calcul le numéro du joueur adversaire
                    ## on crée une nouvelle clé dans le dictionnaire dAdj2 pour ce nouvel état, associée à []
                    dAdj2[next_conf_no] = []
                    ## on examine si la configuration est gagnante pour l'un ou l'autre joueur, et si tel est le cas,
                    ## on l'ajoute à la liste des objectifs du joueur
                    if gagnante(next_conf, joueur):
                        if joueur == 1:
                            LObj12.append(next_conf_no)
                        else:
                            LObj22.append(next_conf_no)
                    ## si la configuration nouvelle n'est pas gagnante, on l'ajoute à la file des configurations
                    ## à examiner afin de déterminer ses successeurs possibles
                    if not gagnante(next_conf, joueur):
                        file.appendleft(next_conf_no)
                ## dans tous les cas (que la configuration soit nouvelle ou non), on doit l'ajouter en tant que
                ## successeur possible de la configuration conf
                dAdj2[conf_no].append(next_conf_no)

t = time() - t0
print(t) ## exemple de temps d'exécution : 9.401000022888184

print(len(confs2), len(LS2), len(dC2), len(dAdj2), len(LObj12), len(LObj22)) ## 5478 5478 5478 5478 626 316
print(len(dAdj2), sum([len(dAdj2[etat]) for etat in dAdj2.keys()])) ## ordre et taille du graphe : (5478, 16167)

9.541999816894531
5478 5478 5478 5478 626 316
5478 16167


* tests

On peut examiner les dix premières configurations construites :

In [109]:
for i in range(11):
    affiche(confs2[i])
    print("*****")

_ _ _
_ _ _
_ _ _
*****
O _ _
_ _ _
_ _ _
*****
_ O _
_ _ _
_ _ _
*****
_ _ O
_ _ _
_ _ _
*****
_ _ _
O _ _
_ _ _
*****
_ _ _
_ O _
_ _ _
*****
_ _ _
_ _ O
_ _ _
*****
_ _ _
_ _ _
O _ _
*****
_ _ _
_ _ _
_ O _
*****
_ _ _
_ _ _
_ _ O
*****
O X _
_ _ _
_ _ _
*****


* avec dictionnaire des configurations

Afin d'accélerer la recherche des configurations dans la liste des configurations déjà construite, on peut maintenir un dictionnaire des configurations déjà trouvées, associées à leur numéro (position dans la liste `confs2`).  
Mais comme les configurations ont été implémentées sous forme d'objets de type `list`, on ne peut les utiliser comme clés d'un dictionnaire.  
Il faut donc se donner une représentation des configurations sous un forme de type non mutable, par exemple : tuples de tuples d'entiers, ou sous forme de chaînes de caractères.  

On fait le choix d'encoder les configurations sous la forme d'une chaîne de caractères de la façon suivante :

In [110]:
def encodage(conf):
    ch = ""
    for i in range(3):
        for j in range(3):
            ch = ch + conf[i][j]
    return ch   

## test
encodage(plateau_vide()), encodage([['O', 'X', 'X'], ['_', '_', '_'], ['O', 'O', '_']])

('_________', 'OXX___OO_')

On reprend l'algorithme précédent, en effectuant les modifications nécessaires.  
On a supprimé les annotations précédentes afin de mettre en évidence uniquement les modifications :

In [111]:
from time import time

t0 = time()

from collections import deque
from copy import deepcopy

## initialisations
confs2 = [plateau_vide()] 
conf_init = 0             
LS2 = [0]                 
dC2 = {0: 1}              
dAdj2 = {0: []}           
LObj12, LObj22 = [], []

##  MODIFICATION ICI
## on initialise le dictionnaire des configurations 
Dconfs2 = {encodage(confs2[0]): 0} ## avec pour seul élément, l'encodage du plateau vide associé à son numéro : 0
## FIN MODIFICATION

## parcours en largeur
file = deque()
file.appendleft(0)
while len(file) > 0:
    conf_no = file.pop()
    conf = confs2[conf_no]
    joueur = dC2[conf_no]
    for i in range(3):
        for j in range(3):
            if lire(conf, i, j) == 0:
                next_conf = deepcopy(conf) ## ou bien next_conf =  [[conf[i][j] for j in range(3)] for i in range(3)]
                jouer(next_conf, i, j, joueur)
                ##  MODIFICATION ICI
                ## on teste à l'aide du dictionnaire Dconfs2 si la configuration a déjà été trouvée
                if encodage(next_conf) in Dconfs2:
                    ## si tel est le cas, on récupère son numéro
                    next_conf_no = Dconfs2[encodage(next_conf)]
                ## sinon, on effectue les traitements nécessaires, comme dans la précédente version de l'algorithme,
                ## à ceci prés qu'il faut ajouter une entrée dans le dictionnaire Dconfs2
                else:
                ## FIN MODIFICATION
                    next_conf_no = len(confs2)
                    confs2.append(next_conf)
                    LS2.append(next_conf_no)
                    ##  MODIFICATION ICI
                    Dconfs2[encodage(next_conf)] = next_conf_no
                    ## FIN MODIFICATION
                    dC2[next_conf_no] = 3 - joueur
                    dAdj2[next_conf_no] = []
                    if gagnante(next_conf, joueur):
                        if joueur == 1:
                            LObj12.append(next_conf_no)
                        else:
                            LObj22.append(next_conf_no)
                    if not gagnante(next_conf, joueur):
                        file.appendleft(next_conf_no)
                dAdj2[conf_no].append(next_conf_no)

t = time() - t0
print(t) ## exemple de temps d'exécution : 0.8780002593994141

print(len(confs2), len(LS2), len(dC2), len(dAdj2), len(LObj12), len(LObj22)) ## 5478 5478 5478 5478 626 316
print(len(dAdj2), sum([len(dAdj2[etat]) for etat in dAdj2.keys()])) ## ordre et taille du graphe : (5478, 16167)

0.618999719619751
5478 5478 5478 5478 626 316
5478 16167


**On peut constater que l'on a gagné un facteur 10 sur le temps d'exécution.**

* test

On peut simuler une partie dans laquelle les deux joueurs jouent aléatoirement à chacun leur tour, un coup parmi tous les coups possibles :

In [112]:
etat = 0
affiche(confs2[etat])
print("\n***** état atteint :", etat)
etats_suivants_possibles = dAdj2[etat]
while len(etats_suivants_possibles) > 0:
    i = randrange(len(etats_suivants_possibles)) # choix du coup suivant
    etat = etats_suivants_possibles[i]
    affiche(confs2[etat])
    print("\n***** état atteint :", etat)
    etats_suivants_possibles = dAdj2[etat]

_ _ _
_ _ _
_ _ _

***** état atteint : 0
_ O _
_ _ _
_ _ _

***** état atteint : 2
_ O _
_ _ _
X _ _

***** état atteint : 23
_ O _
_ _ _
X _ O

***** état atteint : 174
_ O _
_ X _
X _ O

***** état atteint : 617
_ O _
O X _
X _ O

***** état atteint : 1746
_ O _
O X X
X _ O

***** état atteint : 3213
_ O O
O X X
X _ O

***** état atteint : 4659
_ O O
O X X
X X O

***** état atteint : 5301
O O O
O X X
X X O

***** état atteint : 5441


***Il est à noter qu'il serait également possible d'explorer l'arbre des déroulements possibles du jeu à l'aide d'un parcours en profondeur, en utilisant une pile (implémentée ici comme une liste Python) :***

On repris ci-dessous le dernier algorithme, là encore en supprimant les annotations précédentes afin de mettre en évidence uniquement les modifications :

In [113]:
from time import time

t0 = time()

from collections import deque
from copy import deepcopy

## initialisations
confs2 = [plateau_vide()] 
conf_init = 0             
LS2 = [0]                 
dC2 = {0: 1}              
dAdj2 = {0: []}           
LObj12, LObj22 = [], []

Dconfs2 = {encodage(confs2[0]): 0}

## parcours en profondeur
pile = []##  <--- MODIFICATION ICI
pile.append(0)##  <--- MODIFICATION ICI
## FIN MODIFICATION
while len(pile) > 0:##  <--- MODIFICATION ICI
    conf_no = pile.pop()##  <--- MODIFICATION ICI
    conf = confs2[conf_no]
    joueur = dC2[conf_no]
    for i in range(3):
        for j in range(3):
            if lire(conf, i, j) == 0:
                next_conf = deepcopy(conf)
                jouer(next_conf, i, j, joueur)
                if encodage(next_conf) in Dconfs2:
                    next_conf_no = Dconfs2[encodage(next_conf)]
                else:
                    next_conf_no = len(confs2)
                    confs2.append(next_conf)
                    LS2.append(next_conf_no)
                    Dconfs2[encodage(next_conf)] = next_conf_no
                    dC2[next_conf_no] = 3 - joueur
                    dAdj2[next_conf_no] = []
                    if gagnante(next_conf, joueur):
                        if joueur == 1:
                            LObj12.append(next_conf_no)
                        else:
                            LObj22.append(next_conf_no)
                    if not gagnante(next_conf, joueur):
                        pile.append(next_conf_no)##  <--- MODIFICATION ICI
                dAdj2[conf_no].append(next_conf_no)

t = time() - t0
print(t) ## exemple de temps d'exécution : 0.8780002593994141

print(len(confs2), len(LS2), len(dC2), len(dAdj2), len(LObj12), len(LObj22)) ## 5478 5478 5478 5478 626 316
print(len(dAdj2), sum([len(dAdj2[etat]) for etat in dAdj2.keys()])) ## ordre et taille du graphe : (5478, 16167)

0.6590001583099365
5478 5478 5478 5478 626 316
5478 16167


* construction de la représentation du jeu

In [114]:
jeu2 = {'etats': LS2, 'init': 0, 'controles': dC2, 'transitions': dAdj2,
       'objectifs1': LObj12, 'objectifs2': LObj22}

* sauvegarde des résultats sous forme de fichiers texte

On utilise le séparateur tabulation `\t`, et la méthode `.join`, qui permet de concaténer en une unique chaîne des chaînes de caractères stockées dans une liste, en insérant entre elles un séparateur, avec la syntaxe `separateur.join(<liste de chaînes de caractères>)` de la façon suivante :

In [115]:
'**'.join(['a', 'b', 'c', 'd'])

'a**b**c**d'

In [116]:
ch = '\t'.join([str(i) for i in range(1, 16)])
ch

'1\t2\t3\t4\t5\t6\t7\t8\t9\t10\t11\t12\t13\t14\t15'

In [117]:
print(ch)

1	2	3	4	5	6	7	8	9	10	11	12	13	14	15


Ci-dessous la sauvegarde effective :

In [118]:
f = open("confs_LS2.txt", "w")
for no in LS2:
    f.write('\t'.join([s for line in confs2[no] for s in line]) + '\t' + str(no) + '\n')
f.close()

In [119]:
f = open("confs_dC2.txt", "w")
for no, joueur in dC2.items():
    f.write('\t'.join((str(no), str(joueur))) + '\n')
f.close()

In [120]:
f = open("confs_dAdj2.txt", "w")
for etat, Ladj in dAdj2.items():
    #f.write(str(confs[no] + '\t' + str(no) + '\n'))
    f.write('\t'.join([str(etat)] + [str(s) for s in Ladj])+ '\n')
f.close()

In [121]:
f = open("confs_LObj12.txt", "w")
for e in LObj12:
    f.write(str(e) + '\n')
f.close()

In [122]:
f = open("confs_LObj22.txt", "w")
for e in LObj22:
    f.write(str(e) + '\n')
f.close()

On peut obtenir la liste de tous les fichiers annexés au notebook ou créés depuis le début du TD en exécutant la cellule suivante :

In [123]:
import os
os.listdir()

['sauv_dC_morpion.txt', 'sauv_LObj1_morpion.txt', 'sauv_dAdj_morpion.txt', 'sauv_LObj2_morpion.txt', 'sauv_LS_morpion.txt', 'confs_LS1.csv', 'confs_dC1.csv', 'confs_dAdj1.csv', 'confs_LObj11.csv', 'confs_LObj21.csv', 'confs_LS2.txt', 'confs_dC2.txt', 'confs_dAdj2.txt', 'confs_LObj12.txt', 'confs_LObj22.txt']

et on peut les télécharger en décommentant et en exécutant la cellule suivante :

In [124]:
#import basthon
#for f in os.listdir():
#    basthon.download(f)

**Q28. [préliminaire]**

On a mis en annexe de ce notebook une sauvegarde des données sous la forme précédente afin de pouvoir traiter les dernières questions sans avoir traité les parties **3.1** ou **3.2**, on peut charger ces données et obtenir une représentation du jeu en exécutant la cellule suivante :

* chargement des données sauvegardées

In [125]:
f = open("sauv_LS_morpion.txt")
L = f.readlines()
f.close()
confsX, LSX = [''] * len(L), [i for i in range(len(L))]
for line in L:
    line = line.strip().split('\t')
    no = int(line[-1])
    confsX[no] = [[line[3 * i + j] for j in range(3)] for i in range(3)]
len(confsX), len(LSX)

(5478, 5478)

In [126]:
f = open("sauv_dC_morpion.txt")
L = f.readlines()
f.close()
dCX = {}
for line in L:
    etat, joueur = [int(e) for e in line.strip().split('\t')]
    dCX[etat] = joueur
len(dCX)

5478

In [127]:
f = open("sauv_dAdj_morpion.txt")
L = f.readlines()
f.close()
dAdjX = {}
for line in L:
    line = [int(e) for e in line.strip().split('\t')]
    etat, Ladj = line[0], line[1:]
    dAdjX[etat] = Ladj
len(dAdjX), sum([len(dAdjX[etat]) for etat in dAdjX.keys()])

(5478, 16167)

In [128]:
f = open("sauv_LObj1_morpion.txt")
C = f.read()
f.close()
LObj1X = [int(e) for e in C[:-1].split('\n')]
len(LObj1X)

626

In [129]:
f = open("sauv_LObj2_morpion.txt")
C = f.read()
f.close()
LObj2X = [int(e) for e in C[:-1].split('\n')]
len(LObj2X)

316

* restauration d'une représentation du jeu

In [130]:
jeu = {'etats': LSX, 'init': 0, 'controles': dCX, 'transitions': dAdjX,
       'objectifs1': LObj1X, 'objectifs2': LObj2X}

In [131]:
confs = confsX

In [132]:
LS, dC, dAdj, LObj1, LObj2 = LSX, dCX, dAdjX, LObj1X, LObj2X

Pour la suite, on peut se donner un dictionnaire des configurations, dont les clés sont les configurations encodées par la fonction `encodage`, les valeurs associées les numéros d'états correspondants :

In [133]:
Dconfs = {}
for etat in LS:
    Dconfs[encodage(confs[etat])] = etat

On peut alors obtenir le numéro d'état d'une configuration donnée par la fonction suivante :

In [134]:
def get_numero(conf):
    return Dconfs[encodage(conf)]

Par exemple :

In [135]:
conf = [['O', '_', '_'], ['_', '_', '_'], ['_', '_', '_']]
affiche(conf)
get_numero(conf)

O _ _
_ _ _
_ _ _


1

**Q28.**

In [136]:
jeu = jeu1
attr1 = attracteur_infty(jeu, 1, appartientD)
attr2 = attracteur_infty(jeu, 2, appartientD)
len(attr1), len(attr2), len(attr1) + len(attr2), len(jeu['etats'])## 2936 / 1474 / 4410 / 5478

(2936, 1474, 4410, 5478)

In [137]:
jeu = jeu2
attr1 = attracteur_infty(jeu, 1, appartientD)
attr2 = attracteur_infty(jeu, 2, appartientD)
len(attr1), len(attr2), len(attr1) + len(attr2), len(jeu['etats'])## 2936 / 1474

(2936, 1474, 4410, 5478)

In [138]:
jeu = jeuX
attr1 = attracteur_infty(jeu, 1, appartientD)
attr2 = attracteur_infty(jeu, 2, appartientD)
len(attr1), len(attr2), len(attr1) + len(attr2), len(jeu['etats'])## 2936 / 1474

Traceback (most recent call last):
  File "<input>", line 1, in <module>
NameError: name 'jeuX' is not defined


Error: 

* attracteur pour le joueur 1

In [139]:
attr1 = attracteur_infty(jeu2, 1, appartientD)

* attracteur pour le joueur 2

In [140]:
attr2 = attracteur_infty(jeu2, 2, appartientD)

On peut vérifier qu'il existe 2936 états dans l'attracteur du joueur 1, et 1474 dans celui du joueur 2 :

In [141]:
len(attr1), len(attr2)

(2936, 1474)

Ainsi, parmi les 5478 états atteignables du jeu, il y en a $5478-(2936+1474)= 1068$ à partir desquels aucun des deux joueurs ne peut être assuré de gagner.

In [142]:
len(jeu['etats']) - (len(attr1) + len(attr2))

1068

**Q29.**

Il s'agit d'examiner si l'état initial (plateau vide), est au non dans l'attracteur de l'un des deux joueurs :

In [143]:
0 in attr1, 0 in attr2

(False, False)

**On constate qu'il n'existe pas de stratégie gagnante pour aucun des deux joueurs.**

Cependant, la connaissance des attracteurs, permettent d'indiquer, pour une configuration donnée du jeu, les coups à jouer pour s'assurer la victoire, dès qu'il en existe, et les coups à ne pas jouer, pour ne pas la donner à l'adversaire.

Par exemple, pour la configuration suivante :

In [144]:
conf = [['O', '_', '_'], ['_', '_', '_'], ['_', '_', '_']]
affiche(conf)

O _ _
_ _ _
_ _ _


On peut connaître, pour le joueur 2, le seul coup à jouer pour ne pas donner la victoire à l'adversaire :

In [145]:
conf = [['O', '_', '_'], ['_', '_', '_'], ['_', '_', '_']]
etat = get_numero(conf)
for etat_suivant in dAdj[etat]:
    if etat_suivant not in attr1:
        print(etat_suivant)
        affiche(confs[etat_suivant])

5473
O _ _
_ X _
_ _ _


On peut définir une fonction donnant les coups perdants, s'il en existe :

In [146]:
def etats_suivants_perdants(etat):
    joueur = dC[etat]
    if joueur == 1:
        attr_joueur, attr_adversaire = attr1, attr2
    else:
        attr_joueur, attr_adversaire = attr2, attr1
    L = []
    for etat_suivant in dAdj[etat]:
        if etat_suivant in attr_adversaire:
            L.append(etat_suivant)
    return L

In [147]:
conf = [['_', '_', '_'], ['_', '_', '_'], ['_', '_', '_']]
print("configuration de départ :")
affiche(conf)
print("états suivants perdants :")
etat = get_numero(conf)
L = etats_suivants_perdants(etat)
for etat_suivant in L:
    affiche(confs[etat_suivant])

configuration de départ :
_ _ _
_ _ _
_ _ _
états suivants perdants :


In [148]:
conf = [['O', '_', '_'], ['_', '_', '_'], ['_', '_', '_']]
print("configuration de départ :")
affiche(conf)
print("états suivants perdants :")
etat = get_numero(conf)
L = etats_suivants_perdants(etat)
for etat_suivant in L:
    affiche(confs[etat_suivant])

configuration de départ :
O _ _
_ _ _
_ _ _
états suivants perdants :
O X _
_ _ _
_ _ _
O _ X
_ _ _
_ _ _
O _ _
X _ _
_ _ _
O _ _
_ _ X
_ _ _
O _ _
_ _ _
X _ _
O _ _
_ _ _
_ X _
O _ _
_ _ _
_ _ X


De même pour les coups gagnants (ce qui revient à définir une stratégie, mais sans se limiter à un seul choix de coup amenant à la victoire) :

In [149]:
def etats_suivants_gagnants(etat):
    joueur = dC[etat]
    if joueur == 1:
        attr_joueur = attr1
        attr_adversaire = attr2
    else:
        attr_joueur = attr2
        attr_adversaire = attr1
    L = []
    for etat_suivant in dAdj[etat]:
        if etat_suivant in attr_joueur:
            L.append(etat_suivant)
    return L

In [150]:
conf = [['_', '_', '_'], ['_', '_', '_'], ['_', '_', '_']]
print("configuration de départ :")
affiche(conf)
print("états suivants gagnants :")
etat = get_numero(conf)
L = etats_suivants_gagnants(etat)
for etat_suivant in L:
    affiche(confs[etat_suivant])

configuration de départ :
_ _ _
_ _ _
_ _ _
états suivants gagnants :


In [151]:
conf = [['O', '_', '_'], ['_', '_', '_'], ['_', '_', '_']]
print("configuration de départ :")
affiche(conf)
print("états suivants gagnants :")
etat = get_numero(conf)
L = etats_suivants_gagnants(etat)
for etat_suivant in L:
    affiche(confs[etat_suivant])

configuration de départ :
O _ _
_ _ _
_ _ _
états suivants gagnants :


On en déduit enfin une fonction donnant les coups possibles qui sont *&laquo; neutres &raquo;* (ni gagnants ni perdants), ***ce sont les coups qui permettent au joueur de forcer l'adversaire à une partie nulle*** :

In [152]:
def etats_suivants_neutres(etat):
    L = []
    for etat_suivant in dAdj[etat]:
        if etat_suivant not in etats_suivants_perdants(etat) and etat_suivant not in etats_suivants_gagnants(etat):
            L.append(etat_suivant)
    return L

In [153]:
conf = [['_', '_', '_'], ['_', '_', '_'], ['_', '_', '_']]
print("configuration de départ :")
affiche(conf)
print("états suivants neutres :")
etat = get_numero(conf)
L = etats_suivants_neutres(etat)
for etat_suivant in L:
    affiche(confs[etat_suivant])

configuration de départ :
_ _ _
_ _ _
_ _ _
états suivants neutres :
O _ _
_ _ _
_ _ _
_ O _
_ _ _
_ _ _
_ _ O
_ _ _
_ _ _
_ _ _
O _ _
_ _ _
_ _ _
_ O _
_ _ _
_ _ _
_ _ O
_ _ _
_ _ _
_ _ _
O _ _
_ _ _
_ _ _
_ O _
_ _ _
_ _ _
_ _ O


In [154]:
conf = [['O', '_', '_'], ['_', '_', '_'], ['_', '_', '_']]
print("configuration de départ :")
affiche(conf)
print("états suivants neutres :")
etat = get_numero(conf)
L = etats_suivants_neutres(etat)
for etat_suivant in L:
    affiche(confs[etat_suivant])

configuration de départ :
O _ _
_ _ _
_ _ _
états suivants neutres :
O _ _
_ X _
_ _ _


**Q30.**

Il n'existe aucune stratégie gagnante, pour aucun des deux joueurs, mais il est néanmoins utile de faire appel à la fonction `attracteur_infty_strategie`.   

In [155]:
strategie1 = attracteur_infty_strategie(jeu, 1, appartientD)
strategie2 = attracteur_infty_strategie(jeu, 2, appartientD)

In [156]:
len(strategie1), len(strategie2)

(1830, 1006)

Il existe donc 1830 configurations contrôlées par le joueur 1 à partir desquelles il peut s'assurer la victoire, et il en existe 1006 pour le joueur 2.

***Les dictionnaires `strategie1` et `strategie2` donnent, pour chacun des deux joueurs, les coups à jouer pour gagner à coup sûr, à partir des configurations pour lesquelles cela est possible, pour l'un ou l'autre joueur.***

Cependant, **comme l'état initial n'est dans aucun des deux attracteurs,** on en conclut que **si les deux joueurs jouent au mieux de leurs intérêts, toute partie s'achèvera sur une partie nulle**.

On peut simuler une partie et afficher à chaque étape, pour le joueur qui contrôle la configuration, les coups à jouer pour s'assurer la victoire, s'il en existe, et les coups neutres sinon :

***Dans la cellule suivante, passer, sur la première ligne, la variable `simulation`  à `True` puis exécuter la cellule pour lancer une simulation.***

In [160]:
simulation = False
assert simulation

print("On numérote les cases de 1 à 9 (en parcourant la grille de gauche à droite, ligne à après ligne)")
Dstrategies = {1: strategie1, 2: strategie2}
etat = 0
conf_courante = plateau_vide()
affiche(conf_courante)
print("\n********")
while len(dAdj[etat]) > 0:
    joueur = dC[etat]
    print("C'est au joueur", joueur, "de jouer.")
    etats_gagnants = etats_suivants_gagnants(etat)
    print(len(etats_gagnants), "coup(s) gagnant(s)", end=" ;  ")
    etats_perdants = etats_suivants_perdants(etat)
    print(len(etats_perdants), "coup(s) perdant(s)", end=" ;  ")
    etats_neutres = etats_suivants_neutres(etat)
    print(len(etats_neutres), "coup(s) neutre(s)")
    coups_strategie, coups_gagnants, coups_perdants, coups_neutres = [], [], [], []
    nb_vides = 0
    for i in range(3):
        for j in range(3):
            if conf_courante[i][j] == symboles[0]:
                nb_vides += 1
                next_conf = deepcopy(conf_courante)
                jouer(next_conf, i, j, joueur)
                next_conf_no = get_numero(next_conf)
                if etat in Dstrategies[joueur] and next_conf_no == Dstrategies[joueur][etat]:
                    coups_strategie.append(1 + 3 * i + j)
                if next_conf_no in etats_gagnants:
                    coups_gagnants.append(1 + 3 * i + j)
                if next_conf_no in etats_perdants:
                    coups_perdants.append(1 + 3 * i + j)
                if next_conf_no in etats_neutres:
                    coups_neutres.append(1 + 3 * i + j)
    if len(coups_strategie) > 0:
        print("Jouer la case", coups_strategie[0], "(stratégie gagnante)")
    assert len(coups_gagnants) + len(coups_perdants) + len(coups_neutres) == nb_vides
    if len(coups_gagnants) > 0:
        print("Les coups suivants", coups_gagnants, "sont gagnants") 
    if len(coups_perdants) > 0:
        print("Les coups suivants", coups_perdants, "sont perdants") 
    if len(coups_neutres) > 0:
        print("Les coups suivants", coups_neutres, "sont neutres") 
    x = int(input("saisir le numéro de la case à joueur (0 pour arrêter de jouer):\n"))
    if x == 0:
        break
    i, j = (x - 1) // 3, (x - 1) % 3
    jouer(conf_courante, i, j, joueur)
    affiche(conf_courante)
    etat = get_numero(conf_courante)

Traceback (most recent call last):
  File "<input>", line 2, in <module>
AssertionError


Error: 